app/script/backup/BackupRepository.js

```diff
@@ -0,0 +1,157 @@
+    this.backupService = backupService;
+    this.clientRepository = clientRepository;
+    this.userRepository = userRepository;
+    this.ARCHIVE_META_FILENAME = 'meta.json';
+  }
+
+  createMetaDescription() {
+    return {
+      client_id: this.clientRepository.currentClient().id,
+      creation_time: new Date().toISOString(),
+      platform: 'Desktop',
+      user_id: this.userRepository.self().id,
+      version: this.backupService.getDatabaseVersion(),
+    };
+  }
+
+  cancelBackup() {
+    amplify.publish(z.event.WebApp.BACKUP.EXPORT.CANCEL);
+  }
+
+  exportBackup() {
+    return this.backupService.getHistoryCount().then(numberOfRecords => {
+      const userName = this.userRepository.self().username();
+      return {
+        numberOfRecords,
+        userName,
+      };
+    });
+  }
+
+  getUserData() {
+    const clientId = this.clientRepository.currentClient().id;
+    const userId = this.userRepository.self().id;
+
+    return {
+      clientId,
+      userId,
+    };
+  }
+
+  onError(error) {
+    const isBackupImportError = error.constructor.name === 'ImportError';
+
+    amplify.publish(z.event.WebApp.WARNING.MODAL, z.viewModel.ModalsViewModel.TYPE.CONFIRM, {
+      action: () => this.exportBackup(),
+      preventClose: true,
+      text: {
+        action: z.l10n.text(z.string.backupErrorAction),
+        message: isBackupImportError ? z.l10n.text(z.string.backupImportGenericErrorSecondary) : error.message,
+        title: isBackupImportError
+          ? z.l10n.text(z.string.backupImportGenericErrorHeadline)
+          : z.l10n.text(z.string.backupExportGenericErrorHeadline),
+      },
+    });
+  }
+
+  importHistory(archive) {
+    const files = archive.files;
+    if (!files[this.ARCHIVE_META_FILENAME]) {
+      throw new z.backup.InvalidMetaDataError();
+    }
+
+    const metaCheckPromise = files[this.ARCHIVE_META_FILENAME]
+      .async('string')
+      .then(JSON.parse)
+      .then(metadata => checkMetas(metadata, this.createMetaDescription()));
+
+    const unzipPromises = Object.values(archive.files)
+      .filter(zippedFile => zippedFile.name !== this.ARCHIVE_META_FILENAME)
+      .map(zippedFile => zippedFile.async('string').then(value => ({content: value, filename: zippedFile.name})));
+
+    const importEntriesPromise = Promise.all(unzipPromises).then(fileDescriptors => {
+      fileDescriptors.forEach(fileDescriptor => {
+        const tableName = fileDescriptor.filename.replace('.json', '');
+        const entities = JSON.parse(fileDescriptor.content);
+        entities.forEach(entity => this.backupService.importEntity(tableName, entity));
+      });
+    });
+
+    return Promise.all([metaCheckPromise, importEntriesPromise]);
+
+    function checkMetas(archiveMeta, currentMetadata) {
+      if (archiveMeta.user_id !== currentMetadata.user_id) {
+        const message = `History from user "${metadata.user_id}" cannot be restored for user "${user_id}".`;
```

Comment:
user_id is not defined. Must be currentMetadata.user_id

Reply:
Nice catch!